### PR TITLE
Tatau 160

### PR DESCRIPTION
**Pull request by IgorKalaidjan**
Save result after each epoch

### CODE REVIEW

File: tatau_core/sessions/hvd_train.py [hunk 1]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
 
-        ipfs = IPFS()
-        train_weights_path = os.path.join(temp_dir, 'result_weights')
-        model.save_weights(train_weights_path)
+    def _check_task_declaration(self):
+        task_declaration = self.task_assignment.task_declaration.reload()
+        logger.debug('Task Declaration State is {}'.format(task_declaration.state))
+
+        if task_declaration.state == TaskDeclaration.State.CANCELED:
+            raise Exception('This Task Completed by user')
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: Add TaskCanceledExcetion type, for example.

File: tatau_core/sessions/hvd_train.py [hunk 2]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
+            iteration, epoch, accuracy, loss, epoch_time))
+
+        self._check_task_declaration()
+
+        if not epoch % self.nb_iterations:
+            ipfs = IPFS()
+            train_weights_path = os.path.join(self.base_dir, 'result_weights')
+            self.model.save_weights(train_weights_path)
+
+            self.task_assignment.train_result.current_iteration = epoch
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: current_iteration = iteration   not epoch

File: tatau_core/sessions/hvd_train.py [hunk 3]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
+                logger.info('Start evaluation')
+                loss, acc = self.model.eval(chunk_dirs=self.test_chunk_dirs, batch_size=self.batch_size)
+                self.task_assignment.train_result.eval_results[str(iteration)] = {'loss': loss, 'accuracy': acc}
+                logger.info('Finish evaluation')
 
-        with async_commit():
-            self.task_assignment.train_result.save()
+            with async_commit():
+                self.task_assignment.train_result.save()
+                logger.info('Saved epoch: {}'.format(epoch))
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: Saved iteration....

File: tatau_core/sessions/hvd_train.py [hunk 4]
```diff
@@ -288,10 +312,9 @@ def run_mpi(cls):
         session = cls()
         try:
             session.train()
-            exit(0)
         except Exception as ex:
             session.exception = ex
-            logger.debug("Ex: {}".format(ex))
+            logger.debug("Exception in Train: {}".format(ex))
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: please use a single quote instead of double

File: tatau_core/sessions/hvd_train.py [hunk 5]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
-        train_weights_path = os.path.join(temp_dir, 'result_weights')
-        model.save_weights(train_weights_path)
+    def _check_task_declaration(self):
+        task_declaration = self.task_assignment.task_declaration.reload()
+        logger.debug('Task Declaration State is {}'.format(task_declaration.state))
+
+        if task_declaration.state == TaskDeclaration.State.CANCELED:
+            raise Exception('This Task Completed by user')
+
+        return True
```
Comment (Yulya) [inline tatau_core/sessions/hvd_train.py]: remove return True,  we do not expect return value here. We either raise or don't

File: tatau_core/sessions/hvd_train.py [hunk 6]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
+
+    def on_epoch_start(self, epoch):
+        iteration = math.ceil(epoch / self.nb_iterations)
+        logger.info('Progress iteration #{}: start epoch #{}'.format(iteration, epoch))
 
-        self.task_assignment.train_result.weights_ipfs = ipfs.add_file(train_weights_path).multihash
-        self.task_assignment.train_result.loss = train_history['loss'][-1]
-        self.task_assignment.train_result.accuracy = train_history['acc'][-1]
+        self._check_task_declaration()
```
Comment (Yulya) [inline tatau_core/sessions/hvd_train.py]: I suggest move this to the top here

File: tatau_core/sessions/hvd_train.py [hunk 7]
```diff
@@ -37,24 +40,57 @@ def __init__(self, task_assignment_id, uuid=None):
...
 
-        if eval_results:
-            self.task_assignment.train_result.eval_results[str(current_iteration)] = eval_results
+    def on_epoch_end(self, epoch, accuracy, loss, epoch_time, *args, **kwargs):
+        iteration = math.ceil(epoch / self.nb_iterations)
+        logger.info('Progress iteration #{}: end epoch #{}, acc: {}, loss: {}, epoch time: {}'.format(
+            iteration, epoch, accuracy, loss, epoch_time))
+
+        self._check_task_declaration()
```
Comment (Yulya) [inline tatau_core/sessions/hvd_train.py]: and here to the end to save all the progress that we have

File: tatau_core/sessions/hvd_train.py (hunk 6 above)
Comment (IgorKalaidjan) [reply tatau_core/sessions/hvd_train.py]: Done

File: tatau_core/sessions/hvd_train.py (hunk 5 above)
Comment (IgorKalaidjan) [reply tatau_core/sessions/hvd_train.py]: Done

File: tatau_core/sessions/hvd_train.py (hunk 1 above)
Comment (IgorKalaidjan) [reply tatau_core/sessions/hvd_train.py]: Done

File: tatau_core/sessions/hvd_train.py (hunk 6 above)
Comment (Yulya) [reply tatau_core/sessions/hvd_train.py]: > I think, if epoch is finished we should save results and stop work on a task. It's better than losing the last results.

agree, my next comment is about it

File: tatau_core/sessions/hvd_train.py (hunk 3 above)
Comment (IgorKalaidjan) [reply tatau_core/sessions/hvd_train.py]: Done

File: tatau_core/sessions/hvd_train.py (hunk 4 above)
Comment (IgorKalaidjan) [reply tatau_core/sessions/hvd_train.py]: Done

File: tatau_core/sessions/hvd_train.py [hunk 8]
```diff
@@ -37,24 +41,52 @@ def __init__(self, task_assignment_id, uuid=None):
...
 
-        ipfs = IPFS()
-        train_weights_path = os.path.join(temp_dir, 'result_weights')
-        model.save_weights(train_weights_path)
+    def _check_task_declaration(self):
+        task_declaration = self.task_assignment.task_declaration.reload()
+        logger.debug('Task Declaration State is {}'.format(task_declaration.state))
+
+        if task_declaration.state == TaskDeclaration.State.CANCELED:
+            raise TaskCanceledException('This Task Completed by user')
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: this task CANCELED ....

File: tatau_core/sessions/hvd_train.py [hunk 9]
```diff
@@ -37,24 +41,52 @@ def __init__(self, task_assignment_id, uuid=None):
...
+
+        iteration = math.ceil(epoch / self.nb_iterations)
+        logger.info('Progress iteration #{}: start epoch #{}'.format(iteration, epoch))
+
+    def on_epoch_end(self, epoch, accuracy, loss, epoch_time, *args, **kwargs):
+        iteration = math.ceil(epoch / self.nb_iterations)
+        logger.info('Progress iteration #{}: end epoch #{}, acc: {}, loss: {}, epoch time: {}'.format(
+            iteration, epoch, accuracy, loss, epoch_time))
+
+        if not epoch % self.nb_iterations:
```
Comment (makar21) [inline tatau_core/sessions/hvd_train.py]: epoch % epoch_in_iterations ...

**Pull request merged**